---
title: "Cow RNA"
output: html_document
---
```{r setup, include=FALSE}
# knitr::opts_chunk$set(echo = FALSE)
```
A step-by-step workflow for low-level analysis of single-cell RNA-seq data with Bioconductor
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5112579/

Orchestrating Single-Cell Analysis with Bioconductor: Workshop
http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/OSCABioc2019__OSCABioc2019/
https://osca.bioconductor.org/quality-control.html#choice-of-qc-metrics
https://bioconductor.org/packages/devel/bioc/vignettes/scran/inst/doc/scran.html
https://nbisweden.github.io/single-cell_sib_scilifelab/session-dimensionality-reduction/scater_02_dim_reduction_compiled.nb.html#pca
https://bioconductor.org/packages/devel/bioc/vignettes/scater/inst/doc/overview.html

Seurat Tutorial:
https://scrnaseq-course.cog.sanger.ac.uk/website/seurat-chapter.html#setup-the-seurat-object
https://satijalab.org/seurat/v3.2/pbmc3k_tutorial.html
https://learn.gencore.bio.nyu.edu/single-cell-rnaseq/
https://broadinstitute.github.io/2019_scWorkshop/data-wrangling-scrnaseq.html
https://nbisweden.github.io/excelerate-scRNAseq/session-qc/Quality_control.html

https://nbisweden.github.io/workshop-scRNAseq/exercises.html

Current best practices in single‐cell RNA‐seq analysis: a tutorial
https://www.embopress.org/doi/10.15252/msb.20188746

A systematic evaluation of single cell RNA-seq analysis pipelines
https://www.nature.com/articles/s41467-019-12266-7

Single-cell RNA sequencing technologies and bioinformatics pipelines
https://www.nature.com/articles/s12276-018-0071-8#Sec3

Normalizing single-cell RNA sequencing data: challenges and opportunities
https://www.nature.com/articles/nmeth.4292

```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

install.packages("Seurat")
BiocManager::install("scater")
BiocManager::install("scran")
#BiocManager::install("SC3")
```

## Setup
Load required packages
```{r main, message=FALSE}
suppressMessages(require(Seurat))
suppressMessages(require(scater))
```

Read data
```{r}
cow.tpm <- read.table(file="./data/tpmCounts.txt", na.strings = "-", sep = "\t", header = TRUE)
cow.raw <- read.table(file="./data/rawCounts.txt", na.strings = "-", sep = "\t", header = TRUE)
cow.metadata <- read.table(file="./data/annotation.txt", na.strings = "-", sep = "\t", header = TRUE)
cow.gene_name <- read.table(file="./data/geneInfo.txt", na.strings = "-", sep = "\t", header = TRUE)

cow.metadata$Days_Type <- cow.metadata$Days
cow.metadata[cow.metadata$Days != cow.metadata$Cell_type,]$Days_Type <- paste0(cow.metadata[cow.metadata$Days != cow.metadata$Cell_type,]$Days, " - ", cow.metadata[cow.metadata$Days != cow.metadata$Cell_type,]$Cell_type)
cow.metadata$Sample <- NULL #Delete unnecessary column(s) -> reduce file size of exported rds
#saveRDS(cow.metadata, file = "./cow.metadata.rds")
```


Create the Seurat object
```{r}
cow <- CreateSeuratObject(counts = cow.raw, meta.data = cow.metadata)
```

## Quality Control
Calculate mitochondrial proportion and ribosomal proportion
```{r}
mt.genes <- cow.gene_name$Gene.ID[grep("^MT-",cow.gene_name$Gene.Name)]
percent.mito <- Matrix::colSums(cow[mt.genes, ])/Matrix::colSums(cow) * 100
cow <- AddMetaData(cow, percent.mito, col.name = "percent.mito")

rb.genes <- cow.gene_name$Gene.ID[grepl(paste(c("^RP[SL]","5_8S_rRNA","5S_rRNA"), collapse = "|"), cow.gene_name$Gene.Name)]
percent.ribo <- Matrix::colSums(cow[rb.genes, ])/Matrix::colSums(cow) * 100
cow <- AddMetaData(cow, percent.ribo, col.name = "percent.ribo")
```

Create the Single Cell Experiment object
```{r}
sce <- as.SingleCellExperiment(cow)
sce <- addPerCellQC(sce, subsets = list(mito = mt.genes))
sce <- addPerFeatureQC(sce)
```

### Plot QC {.tabset}
#### Histograms
```{r}
hist(sce$total/1e6, xlab="Library sizes(Milions)", main="",breaks=50, col="grey80", ylab="Number of cells")
abline(v="0.7",col="red")

hist(sce$detected, xlab="Number of expressed genes", main="",breaks=50, col="grey80", ylab="Number of cells")

hist(sce$subsets_mito_percent, xlab="MT %", main="",breaks=50, col="grey80", ylab="Number of cells")
abline(v="20",col="red")
```

#### VlnPlot
More than 3 MADs = Outlier
```{r}
#libsize.drop <- isOutlier(sce$total, nmads=3, type="lower", log=TRUE)
#feature.drop <- isOutlier(sce$detected, nmads=3, type="lower", log=TRUE)
#mito.drop <- isOutlier(sce$subsets_mito_percent, nmads=3, type="higher")

#attr(libsize.drop, "thresholds")
#attr(feature.drop, "thresholds")
#attr(mito.drop, "thresholds")

#sce_filtered <- sce[,!(libsize.drop | feature.drop | mito.drop)]
#data.frame(ByLibSize=sum(libsize.drop), ByLibSizeUpper=sum(libsize_u.drop), ByFeature=sum(feature.drop),ByMito=sum(mito.drop), Remaining=ncol(sce_filtered))

reasons <- quickPerCellQC(sce, percent_subsets=c("subsets_mito_percent"), batch=sce@colData@listData[["Experiment"]])

#attr(reasons@listData[["low_lib_size"]], "thresholds")
#attr(reasons@listData[["low_n_features"]], "thresholds")
#attr(reasons@listData[["high_subsets_mito_percent"]], "thresholds")

reasons@listData[["low_lib_size"]] <- I(reasons@listData[["low_lib_size"]] | sce$nCount_RNA > 0.7 * 1e6)
reasons@listData[["high_subsets_mito_percent"]] <- I(reasons@listData[["high_subsets_mito_percent"]] | sce$percent.mito > 20)

plotColData(sce, x="Experiment", y="total", colour_by=reasons@listData[["low_lib_size"]]) + scale_y_log10()
plotColData(sce, x="Experiment", y="detected", colour_by=I(reasons@listData[["low_n_features"]])) + scale_y_log10()
plotColData(sce, x="Experiment", y="percent.mito", colour_by=reasons@listData[["high_subsets_mito_percent"]])
```
```{r, include=FALSE, eval=FALSE}
sce$discard <- reasons$discard
gridExtra::grid.arrange(
    plotColData(sce, x="Experiment", y="sum", colour_by="discard")  + 
        scale_y_log10() + ggtitle("Total count"),
    plotColData(sce, x="Experiment", y="detected", colour_by="discard")  + 
        scale_y_log10() + ggtitle("Detected features"),
    plotColData(sce, x="Experiment", y="subsets_mito_percent", 
        colour_by="discard") + 
        ggtitle("Mito percent"),
    ncol=1
)
```

#### Scatter (Raw)
```{r}
plotColData(sce, x="nCount_RNA", y="nFeature_RNA", colour_by="Experiment") + scale_x_log10()
plotColData(sce, x="nFeature_RNA", y="percent.mito", colour_by="Experiment") 
plotColData(sce, x="nCount_RNA", y="percent.mito", colour_by="Experiment") + scale_x_log10()
```

#### Filtering
```{r}
#cow_filtered <- as.Seurat(sce_filtered) #subset(cow, subset = nCount_RNA > 755 & nFeature_RNA > 2699 & percent.mito < 24)
sce_filtered <- sce[,!(reasons$discard | sce$nCount_RNA > 0.7*1e6 | sce$percent.mito > 20)]

#Remove Y chr features
sce_filtered <- sce_filtered[!(row.names(sce_filtered) %in% cow.gene_name$Gene.Name[cow.gene_name$Reference == "NC_016145.1"])]

#Filter out features that are not expressed in any cells:
summary(nexprs(sce_filtered, byrow=TRUE))
sce_filtered <- sce_filtered[nexprs(sce_filtered, byrow=TRUE) > 0,]

cow.tpm <- subset(cow.tpm, rownames(cow.tpm) %in% rownames(sce_filtered))
```

Rename
```{r}
gene_info <- cow.gene_name[match(rownames(sce_filtered), cow.gene_name$Gene.ID),]
row.names(gene_info) <- gene_info$Gene.ID <- row.names(sce_filtered)
gene_info$Gene.Name[is.na(gene_info$Gene.Name)] <- gene_info$Gene.ID[is.na(gene_info$Gene.Name)]

gene_info$otherRNAs <- ifelse(
  gene_info$Gene.ID %in% rb.genes |
  grepl(paste(c("^U[0-9]{1,2}$","^SNOR","U6atac","U4atac"), collapse = "|"), gene_info$Gene.Name), TRUE, FALSE)

#dup_conditions <- gene_info_rename$Gene.Name %in% gene_info_rename$Gene.Name[duplicated(gene_info_rename$Gene.Name)]
#gene_info_rename$Gene.Name[dup_conditions] <- gene_info_rename$Gene.ID[dup_conditions]
gene_info$Gene.AltName <- make.unique(as.character(gene_info$Gene.Name), sep = "*")
gene_info$Gene.AltName <- gsub('_', '-', gene_info$Gene.AltName)

sub_gene_info <- as.data.frame(cbind(Gene.ID=gene_info$Gene.ID,Gene.Name=gene_info$Gene.Name))
rownames(sub_gene_info) <- gene_info$Gene.AltName

rownames(sce_filtered) <- rownames(sub_gene_info)
rownames(cow.tpm) <- rownames(sub_gene_info)
cow_filtered <- as.Seurat(sce_filtered)

cow_tpm <- CreateSeuratObject(counts = cow.tpm, meta.data = cow.metadata)
cow_tpm[["RNA"]]@data <- as(log(cow_tpm[["RNA"]]@counts+1), "dgCMatrix")

saveRDS(sub_gene_info, file = "./gene_info.rds")
saveRDS(cow_filtered, file = "./cow_filtered.rds")
saveRDS(sce_filtered, file = "./sce_filtered.rds")
saveRDS(cow_tpm, file = "./cow_tpm.rds")

#int_elementMetadata(sce_filtered) <- DataFrame(gene_info)
#cow[["RNA"]] <- AddMetaData(cow[["RNA"]], name_map$Gene.Name, "Gene.Name")
#cow[["RNA"]] <- AddMetaData(cow[["RNA"]], name_map$Gene.ID, "Gene.ID")
```

Annotation
```{r}
#Bovine bEPI Preimplantation TOPModular Genes
bEPI <- c('GDF3','PRDM14','TDGF1','FGF4','ZIC3','NANOG','SOX15','LIN28B','ZFP42', 'IFITM1', 'IFITM3','TDF1P3','PRICKLE1','DPPA5','KLF17','KLF4','ARGFX','ESRG','MRS2','POU5F1', 'LEFTY2', 'WNT3','VENTX','SERINC5','FBP1','MT1X','ATG3','CDHR1','DND1','SAT1','PARP1','CFLAR','MAN1C1','CD9','CAPG','SOX2','PIM2','TVC1D23','UNC5B','DPPA2','MEG3','ASH2L','MSH6','CBFA2T2','MRPS23','USP28','BCOR','VCAN','ETV4','CNIH4','DEPTOR','ABHD12B','NODAL','GPR160','SPRY2','SLC39A10','WARS','ASRGL1',"ESRRB")

#Bovineb PE New Preimplantation TOP Modular Genes
bPE <- c('UACA','PDGFRA', 'GATA6', 'GATA4', 'COL4A1', 'HNF1B', 'NID2', 'RSPO3', 'APOA1','SOX17','GAPDA', 'FN1', 'LAMB1','LBH', 'KIT', 'FGFR2','GPX2', 'LAMA4', 'LAMA1','BMP2','SERPINH1','P4HA1','EGLN3','ZC3HAV1','BAMBI','GPRC5B', 'MARCKS','HNF4A', 'DUSP1','ALDH2','APOC1','TCEA1','CDC42EP4','PHLDA1','SEPT11','ENO2','SLC4A8','HORMAD2-AS1','ST3GAL1','NDUFAB1','SPARC','SPATS2L','TMBIM1','UQCRH','TPST2','GYPC','RCBTB1','PXDN','TBCA','RAB15','CLDN19','KLHL18','CADM1','FGFR1','UBASH3B','SRGAP1','GSN','CTSE','GSTO1','BRDT','OTX2','LARP6','UQCRHL')

#Bovine bTE New Preimplantation TOP Modular Genes
bTE <- c('CLDN4','GATA3','GATA2','KRT18','PPIA','CDX2','TEAD4','DAB2','SLC7A2','ABCG2', 'MYC', 'LRP2', 'WNT7A', 'FDGFA', 'FRDM4','KRT8','GRHL2','TACSTD2','MPZL1','PALLD','TACC1','LRRFIP1','RAB11FIP4','RALBP1','PTGES','EMP2','PTN','SH3KBP1','SH2D4A','TEAD1','MGST3','TGFBR3','ODC1','S100A6','JUP','TCF7L2','PRSS8','CEBPA','SLC7A4','ZFHX3','VAMP8','ENTPD1','HIC2','SLC7A5','DLX3','EFNA1','TIGAR','GRHL1','FOLR1','CD55','GAB2','ADK','PPT1','PERP','FHL2','KRT19','TMEM106C',"TMEM54")

sub_gene_info$Annotation <- "UN" 
sub_gene_info[sub_gene_info$Gene.Name %in% bEPI,]$Annotation <- "EPI"
sub_gene_info[sub_gene_info$Gene.Name %in% bPE,]$Annotation <- "PE"
sub_gene_info[sub_gene_info$Gene.Name %in% bTE,]$Annotation <- "TE"

saveRDS(sub_gene_info, file = "./gene_info.rds")
```


#### VlnPlot & Scatter (Filtered)
```{r}
plotColData(sce_filtered, x="Experiment", y="total") + scale_y_log10()
plotColData(sce_filtered, x="Experiment", y="detected") + scale_y_log10()
plotColData(sce_filtered, x="Experiment", y="percent.mito")

plotColData(sce_filtered, x="nCount_RNA", y="nFeature_RNA", colour_by="Experiment") + scale_x_log10()
plotColData(sce_filtered, x="nFeature_RNA", y="percent.mito", colour_by="Experiment") 
plotColData(sce_filtered, x="nCount_RNA", y="percent.mito", colour_by="Experiment") + scale_x_log10()
```

## Normalization {.tabset}

### Seruat LogNormalize
```{r}
cow_logNorm2 <- NormalizeData(object = cow_filtered, normalization.method = "LogNormalize", scale.factor = 10000)
cow_logNorm <- FindVariableFeatures(cow_logNorm, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(cow_logNorm)
cow_logNorm <- ScaleData(object = cow_logNorm, features = all.genes)

saveRDS(cow_logNorm, file = "./cow_logNorm.rds")
```

### Sctransform
```{r, eval=FALSE, message=FALSE, warning=FALSE}
cow_sct <- suppressMessages(SCTransform(cow_filtered, verbose = FALSE, variable.features.n = 2000, return.only.var.genes = FALSE))
#summary(nexprs(cow_sct@assays$RNA@counts[!(rownames(cow_sct@assays$RNA@counts) %in% rownames(cow_sct@assays$SCT@counts) ),], byrow=TRUE))
saveRDS(cow_sct, file = "./cow_sct.rds")
```

### Scran
```{r}
library(scran)

set.seed(100)
clusters <- quickCluster(sce_filtered)
sce_filtered <- computeSumFactors(sce_filtered, clusters=clusters)  

sce_scran <- logNormCounts(sce_filtered)
sce_scran_noLog <- logNormCounts(sce_filtered, log = FALSE) # without(!) log transform

saveRDS(sce_scran, file = "./sce_scran.rds")
saveRDS(sce_scran_noLog, file = "./sce_scran_noLog.rds")

cow_scran <- cow_filtered
cow_scran[["RNA"]]@data <- as(log(assay(sce_scran_noLog, "normcounts")+1), "dgCMatrix")

cow_scran <- FindVariableFeatures(cow_scran, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(cow_scran)
cow_scran <- ScaleData(object = cow_scran, features = all.genes)

saveRDS(cow_scran, file = "./cow_scran.rds")
```

## PCA {.tabset}

### (Seurat Log Normalize)
```{r}
cow_logNorm <- RunPCA(cow_logNorm, npcs = 30, verbose = FALSE)

## Determine PC
cow_logNorm_temp <- JackStraw(cow_logNorm, num.replicate = 100)
cow_logNorm_temp <- ScoreJackStraw(cow_logNorm_temp, dims = 1:20)
JackStrawPlot(cow_logNorm_temp, dims = 1:20)
ElbowPlot(cow_logNorm_temp)

set.seed(100)
cow_logNorm <- RunUMAP(cow_logNorm, reduction = "pca", dims = 1:20, verbose = FALSE, n.components=4)
cow_logNorm <- RunTSNE(cow_logNorm, reduction = "pca", dims = 1:20, verbose = FALSE, dim.embed=3, perplexity=30)

DimPlot(cow_logNorm, reduction = "pca", group.by = "Experiment")
DimPlot(cow_logNorm, reduction = "umap", group.by = "Experiment")
DimPlot(cow_logNorm, reduction = "umap", group.by = "Cell_type")
DimPlot(cow_logNorm, reduction = "tsne", group.by = "Experiment")
DimPlot(cow_logNorm, reduction = "tsne", group.by = "Cell_type")

saveRDS(cow_logNorm, file = "./cow_logNorm.rds")
```

### (Seurat SCT)
```{r, eval = FALSE}
cow_sct <- RunPCA(cow_sct, npcs = 30, verbose = FALSE)

## Determine PC
ElbowPlot(cow_sct)

set.seed(100)
cow_sct <- RunUMAP(cow_sct, reduction = "pca", dims = 1:21, verbose = FALSE, n.components=4)
cow_sct <- RunTSNE(cow_sct, reduction = "pca", dims = 1:21, verbose = FALSE, dim.embed=3, perplexity=30)

DimPlot(cow_sct, reduction = "pca", group.by = "Experiment")
DimPlot(cow_sct, reduction = "umap", group.by = "Experiment")
DimPlot(cow_sct, reduction = "umap", group.by = "Cell_type")
DimPlot(cow_sct, reduction = "tsne", group.by = "Experiment")
DimPlot(cow_sct, reduction = "tsne", group.by = "Cell_type")

saveRDS(cow_sct, file = "./cow_sct.rds")
```
### (Scran - PCA by Seurat)
```{r}
cow_scran <- RunPCA(cow_scran, npcs = 30, verbose = FALSE)

## Determine PC
cow_scran_temp <- JackStraw(cow_scran, num.replicate = 100)
cow_scran_temp <- ScoreJackStraw(cow_scran_temp, dims = 1:20)
JackStrawPlot(cow_scran_temp, dims = 1:20)
ElbowPlot(cow_scran_temp)

set.seed(100)

cow_scran <- RunUMAP(cow_scran, reduction = "pca", dims = 1:19, verbose = FALSE, n.components=4)
cow_scran <- RunTSNE(cow_scran, reduction = "pca", dims = 1:19, verbose = FALSE, dim.embed=3, perplexity=30)

DimPlot(cow_scran, reduction = "pca", group.by = "Experiment")
DimPlot(cow_scran, reduction = "umap", group.by = "Experiment")
DimPlot(cow_scran, reduction = "umap", group.by = "Cell_type")
DimPlot(cow_scran, reduction = "tsne", group.by = "Experiment")
DimPlot(cow_scran, reduction = "tsne", group.by = "Cell_type")

saveRDS(cow_scran, file = "./cow_scran.rds")
```
### (Scran)
```{r}
dec <- modelGeneVar(sce_scran)
top.hvgs <- getTopHVGs(dec, n=2000)
sce_scran <- scater::runPCA(sce_scran, ncomponents = 30, subset_row = top.hvgs, scale = TRUE)
```

Determine no. of PC
```{r}
#1) Using elbow plot
plot(attr(reducedDim(sce_scran, 'PCA'), 'percentVar'))

#2) Using the technical noise
denoised.scran <- denoisePCA(sce_scran, technical=dec, subset.row=top.hvgs)
ncol(reducedDim(denoised.scran))

#3) Based on population structure
choices  <- getClusteredPCs(reducedDim(sce_scran, 'PCA'))
metadata(choices)$chosen 
#reducedDim(sce_scran, "PCAsub") <- reducedDim(sce_scran, "PCA")[,1:npcs,drop=FALSE]
```


```{r}
set.seed(100)
sce_scran <- scater::runUMAP(sce_scran, dimred = 'PCA', n_dimred = 7, ncomponents = 4)
sce_scran <- scater::runTSNE(sce_scran, use_dimred = "PCA", n_dimred = 7, ncomponents = 3, perplexity=30)

plotReducedDim(sce_scran, dimred="PCA", colour_by="Experiment")
plotReducedDim(sce_scran, dimred="UMAP", colour_by="Experiment")
plotReducedDim(sce_scran, dimred="UMAP", colour_by="Cell_type")
plotReducedDim(sce_scran, dimred="TSNE", colour_by="Experiment")
plotReducedDim(sce_scran, dimred="TSNE", colour_by="Cell_type")

saveRDS(sce_scran, file = "./sce_scran.rds")
```
### (TPM)
```{r}
cow_tpm <- FindVariableFeatures(cow_tpm, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(cow_tpm)
cow_tpm <- ScaleData(object = cow_tpm, features = all.genes)
cow_tpm <- RunPCA(cow_tpm, npcs = 30, verbose = FALSE)

## Determine PC
cow_tpm_temp <- JackStraw(cow_tpm, num.replicate = 100)
cow_tpm_temp <- ScoreJackStraw(cow_tpm_temp, dims = 1:20)
JackStrawPlot(cow_tpm_temp, dims = 1:20)
ElbowPlot(cow_tpm_temp)

set.seed(100)

cow_tpm <- RunUMAP(cow_tpm, reduction = "pca", dims = 1:12, verbose = FALSE, n.components=4)
cow_tpm <- RunTSNE(cow_tpm, reduction = "pca", dims = 1:12, verbose = FALSE, dim.embed=3, perplexity=30)


DimPlot(cow_tpm, reduction = "pca", group.by = "Experiment")
DimPlot(cow_tpm, reduction = "umap", group.by = "Experiment")
DimPlot(cow_tpm, reduction = "umap", group.by = "Cell_type")
DimPlot(cow_tpm, reduction = "tsne", group.by = "Experiment")
DimPlot(cow_tpm, reduction = "tsne", group.by = "Cell_type")

saveRDS(cow_tpm, file = "./cow_tpm.rds")
```

## Clustering
```{r}
# library(SC3)
# cow_logNorm.sce <- as.SingleCellExperiment(cow_logNorm)
# cow_sct.sce <- as.SingleCellExperiment(cow_sct)
# cow_scran.sce <- as.SingleCellExperiment(cow_scran)
# cow_tpm.sce <- as.SingleCellExperiment(cow_tpm)
# 
# rowData(cow_logNorm.sce)$feature_symbol <- rownames(cow_logNorm.sce)
# counts(cow_logNorm.sce) <- as.matrix(counts(cow_logNorm.sce))
# logcounts(cow_logNorm.sce) <- as.matrix(logcounts(cow_logNorm.sce))
# cow_logNorm.sce <- sc3(cow_logNorm.sce, ks= 6:8, n_cores = 1) #biology = TRUE, k_estimator = TRUE
# saveRDS(cow_logNorm.sce, file = "./cow_logNorm.sce.rds")
# 
# rowData(cow_sct.sce)$feature_symbol <- rownames(cow_sct.sce)
# counts(cow_sct.sce) <- as.matrix(counts(cow_sct.sce))
# logcounts(cow_sct.sce) <- as.matrix(logcounts(cow_sct.sce))
# cow_sct.sce <- sc3(cow_sct.sce, ks= 6:8, n_cores = 1) #biology = TRUE, k_estimator = TRUE
# saveRDS(cow_sct.sce, file = "./cow_sct.sce.rds")
# 
# rowData(cow_scran.sce)$feature_symbol <- rownames(cow_scran.sce)
# counts(cow_scran.sce) <- as.matrix(counts(cow_scran.sce))
# logcounts(cow_scran.sce) <- as.matrix(logcounts(cow_scran.sce))
# cow_scran.sce <- sc3(cow_scran.sce, ks= 6:8, n_cores = 1) #biology = TRUE, k_estimator = TRUE
# saveRDS(cow_scran.sce, file = "./cow_scran.sce.rds")
# 
# rowData(cow_tpm.sce)$feature_symbol <- rownames(cow_tpm.sce)
# counts(cow_tpm.sce) <- as.matrix(counts(cow_tpm.sce))
# logcounts(cow_tpm.sce) <- as.matrix(logcounts(cow_tpm.sce))
# cow_tpm.sce <- sc3(cow_tpm.sce, ks= 6:8, n_cores = 1) #biology = TRUE, k_estimator = TRUE
# saveRDS(cow_tpm.sce, file = "./cow_tpm.sce.rds")

cow_sct <- FindNeighbors(cow_sct, dims = 1:21)
cow_sct <- FindClusters(cow_sct, resolution = 0.5)
saveRDS(cow_sct, file = "./cow_sct.rds")

cow_tpm <- FindNeighbors(cow_tpm, dims = 1:12)
cow_tpm <- FindClusters(cow_tpm, resolution = 0.5)
saveRDS(cow_tpm, file = "./cow_tpm.rds")
```

## DE
```{r}
de_tpm <- FindAllMarkers(cow_tpm, min.pct = 0.25, logfc.threshold = 0.5)
de_sct <- FindAllMarkers(cow_sct, min.pct = 0.25, logfc.threshold = 0.5)
saveRDS(de_tpm, file = "./de_tpm.rds")
saveRDS(de_sct, file = "./de_sct.rds")
```
```{r}
cow_sct.sce <- as.SingleCellExperiment(cow_sct)
cow_tpm.sce <- as.SingleCellExperiment(cow_tpm)
saveRDS(cow_sct.sce, file = "./cow_sct.sce.rds")
saveRDS(cow_tpm.sce, file = "./cow_tpm.sce.rds")
```


## Integration {.tabset}
### (Seurat Log Normalize)
```{r}
cow_logNorm.list <- SplitObject(cow_logNorm, split.by = "Experiment")

for (i in 1:length(cow_logNorm.list)) {
    cow_logNorm.list[[i]] <- NormalizeData(cow_logNorm.list[[i]], verbose = FALSE)
    #cow_logNorm.list[[i]] <- FindVariableFeatures(cow_logNorm.list[[i]], selection.method = "vst", nfeatures = 2000, verbose = FALSE)
}

reference.list <- cow_logNorm.list[c("Exp1", "Exp2", "Exp3")]
k.filter <- min(200, sapply(cow_logNorm.list, ncol)) 
cow_logNorm.anchors <- FindIntegrationAnchors(object.list = reference.list, dims = 1:30, k.filter = k.filter)

#all_genes <- row.names(cow_logNorm)
cow_logNorm.integrated <- IntegrateData(anchorset = cow_logNorm.anchors, dims = 1:30) #, features.to.integrate =  all_genes)

DefaultAssay(cow_logNorm.integrated) <- "integrated"
cow_logNorm.integrated <- ScaleData(cow_logNorm.integrated, verbose = FALSE)
cow_logNorm.integrated <- RunPCA(cow_logNorm.integrated, npcs = 30, verbose = FALSE)
cow_logNorm.integrated <- RunUMAP(cow_logNorm.integrated, reduction = "pca", dims = 1:30)

DimPlot(cow_logNorm.integrated, reduction = "pca", group.by = "Experiment")

DimPlot(cow_logNorm.integrated, reduction = "umap", group.by = "Experiment")
DimPlot(cow_logNorm.integrated, reduction = "umap", group.by = "Cell_type", label = TRUE, repel = TRUE) + NoLegend()
```

### (Seurat SCT)
```{r, eval=FALSE}
cow_sct.list <- SplitObject(cow_filtered, split.by = "Experiment")

cow_sct.list <- cow_sct.list[c("Exp1", "Exp2", "Exp3")]
for (i in 1:length(cow_sct.list)) {
    cow_sct.list[[i]] <- SCTransform(cow_sct.list[[i]], verbose = FALSE)
}

saveRDS(cow_sct, file = "./cow_sct.rds")

cow_sct.features <- SelectIntegrationFeatures(object.list = cow_sct.list, nfeatures = 3000)
cow_sct.list <- PrepSCTIntegration(object.list = cow_sct.list, anchor.features = cow_sct.features, 
    verbose = FALSE)

cow_sct.anchors <- FindIntegrationAnchors(object.list = cow_sct.list, normalization.method = "SCT", 
    anchor.features = cow_sct.features, verbose = FALSE)
cow_sct.integrated <- IntegrateData(anchorset = cow_sct.anchors, normalization.method = "SCT", 
    verbose = FALSE)

cow_sct.integrated <- RunPCA(cow_sct.integrated, npcs = 30, verbose = FALSE)
cow_sct.integrated <- RunUMAP(cow_sct.integrated, reduction = "pca", dims = 1:30)

DimPlot(cow_sct.integrated, reduction = "pca", group.by = "Experiment")

DimPlot(cow_sct.integrated, reduction = "umap", group.by = "Experiment")
DimPlot(cow_sct.integrated, reduction = "umap", group.by = "Cell_type", label = TRUE, repel = TRUE) + NoLegend()
```

### (Scran Fastmnn)
```{r}
library(batchelor)
  
#rescaled <- rescaleBatches(sce_scran, batch=sce_scran$Experiment)
sce_scran.fastmnn <- sce_scran
sce_scran.fastmnn <- fastMNN(sce_scran.fastmnn, batch=sce_scran.fastmnn$Experiment)
reducedDim(sce_scran.fastmnn, "MNN") <- reducedDim(sce_scran.fastmnn, "corrected")

sce_scran.fastmnn <-  scater::runPCA(sce_scran.fastmnn, dimred = 'MNN', ncomponents = 30)
sce_scran.fastmnn <-  scater::runUMAP(sce_scran.fastmnn, dimred = 'PCA', n_dimred = 30, ncomponents = 30)
sce_scran.fastmnn <-  scater::runTSNE(sce_scran.fastmnn, use_dimred = "PCA", n_dimred = 30, ncomponents = 3)

plotReducedDim(sce_scran.fastmnn, dimred="PCA", colour_by="Experiment")
plotReducedDim(sce_scran.fastmnn, dimred="UMAP", colour_by="Experiment")
plotReducedDim(sce_scran.fastmnn, dimred="TSNE", colour_by="Experiment")
```

###(Scran Liger)
```{r}
cow_scran <- as.Seurat(sce_scran)

library(liger)
library(SeuratData)
library(SeuratWrappers)

cow_scran <- ScaleData(cow_scran, split.by = "Experiment", do.center = FALSE)
cow_scran <- RunOptimizeALS(cow_scran, k = 20, lambda = 5, split.by = "Experiment")
cow_scran <- RunQuantileNorm(cow_scran, split.by = "Experiment")


#cow_scran <-  RunUMAP(cow_scran, dims = 1:ncol(pbmcsca[["iNMF"]]), reduction = "iNMF")
#
#DimPlot(cow_logNorm, reduction = "pca", group.by = "Experiment")
#DimPlot(cow_logNorm, reduction = "umap", group.by = "Experiment")


sce_scran.liger <-  as.SingleCellExperiment(cow_scran)

sce_scran.liger <-  scater::runPCA(sce_scran.liger, dimred = 'INMF', ncomponents = 9)
sce_scran.liger <- scater::runUMAP(sce_scran.liger, dimred = 'PCA', ncomponents = 9)
sce_scran.liger <- scater::runTSNE(sce_scran.liger, use_dimred = "PCA", ncomponents = 3)

plotReducedDim(sce_scran.liger, dimred="PCA", colour_by="Experiment")
plotReducedDim(sce_scran.liger, dimred="UMAP", colour_by="Experiment")
plotReducedDim(sce_scran.liger, dimred="TSNE", colour_by="Experiment")

saveRDS(sce_scran.liger, file = "./sce_scran.liger.rds")
```
